synapse/handlers/sso.py

```diff
@@ -701,8 +707,69 @@ async def _register_mapped_user(
         await self._store.record_user_external_id(
             auth_provider_id, remote_user_id, registered_user_id
         )
+
+        # Set avatar, if available
+        if attributes.picture:
+            await self.set_avatar(registered_user_id, attributes.picture)
+
         return registered_user_id
 
+    async def set_avatar(self, user_id: str, picture_https_url: str) -> None:
+        try:
+            uid = UserID.from_string(user_id)
+
+            # HEAD request to find image size & mime type before download
+            response = await self._http_client.request("HEAD", picture_https_url)
+            if response.code != 200:
+                raise Exception("error sending HEAD request to get image size")
+
+            content_length = int(
+                response.headers.getRawHeaders(b"Content-Length")[0].decode("utf-8")
+            )
+            content_type = response.headers.getRawHeaders(b"Content-Type")[0].decode(
+                "utf-8"
+            )
+
+            # ensure picture size respects max_avatar_size defined in config
+            if self._profile_handler.max_avatar_size is not None:
+                if content_length > self._profile_handler.max_avatar_size:
+                    raise Exception("sso avatar too big in size")
+
+            # ensure picture is of allowed mime type
+            if (
+                self._profile_handler.allowed_avatar_mimetypes
+                and content_type not in self._profile_handler.allowed_avatar_mimetypes
+            ):
+                raise Exception("mime type not allowed for sso avatar")
+
+            # download picture
+            response = await self._http_client.request("GET", picture_https_url)
+            if response.code != 200:
+                raise Exception(
+                    "error sending GET request to provided sso avatar image"
+                )
```

Comment:
Is there anything we can include from the response in the error message to help debug this?

Reply:
Improved all exception messages. Should I add url and matrix user id in the exception messages too?